aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,128 @@
+/**
+ * @file aesd-circular-buffer.c
+ * @brief Functions and data related to a circular buffer imlementation
+ *
+ * @author Dan Walkes
+ * @date 2020-03-01
+ * @copyright Copyright (c) 2020
+ *
+ */
+
+#ifdef __KERNEL__
+#include <linux/string.h>
+#else
+#include <string.h>
+#include <stdio.h>
+#endif
+
+#include "aesd-circular-buffer.h"
+
+/**
+ * @param buffer the buffer to search for corresponding offset.  Any necessary locking must be performed by caller.
+ * @param char_offset the position to search for in the buffer list, describing the zero referenced
+ *      character index if all buffer strings were concatenated end to end
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer.
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+                                                                          size_t char_offset, size_t *entry_offset_byte_rtn)
+{
+    if (buffer == NULL || entry_offset_byte_rtn == NULL)
+    {
+        return NULL;
+    }
+
+    int curr_entry = buffer->out_offs;
+    size_t entry_len;
+    do
+    {
+        entry_len = buffer->entry[curr_entry].size;
+        if (entry_len == 0)
+        {
+            return NULL;
+        }
+        else if (char_offset == 0)
+        {
+            *entry_offset_byte_rtn = 0;
+            return &(buffer->entry[curr_entry]);
+        }
+        else if (entry_len > char_offset)
+        {
+            *entry_offset_byte_rtn = char_offset;
+            return &(buffer->entry[curr_entry]);
```

Comment:
@zeog1488 I think you could combine this else if with the previous one without checking for char_offset = 0 and just assigning it.